{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 1 Lab: next generation sequencing and mutation hunting\n",
    "\n",
    "## Part 2\n",
    "**Due: Monday 4/8/19**\n",
    "\n",
    "**Partner: ** (optional. Indicate which partner is turning in the assignment)\n",
    "\n",
    "Skills covered: TODO\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Set up code\n",
    "%pylab inline\n",
    "import os\n",
    "import numpy as np\n",
    "\n",
    "HOMEDIR=\"/home/mgymrek/workspace/cse185-spring19/1_NGS/data_prep\" # TODO change"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Preparing for SNP calling: pileup files\n",
    "\n",
    "The goal now is to go through our data, and for each position in the reference genome, see how many reads have a mutation at the same position. The tview image looks great, but would take too long to go through the entire human genome. The SAM file is inconvenient for this, because the reads aren’t matched with the reference in a simple way (though they do contain all the information needed to make the alignment in tview). The solution is to make an intermediate file type called an mpileup, because it goes through each position and “piles up” the reads, tabulating the number of bases that match or don’t match the reference.\n",
    "\n",
    "Like tview, mpileup requires a sorted, indexed bam file. We will focus only on a specific region of the genome to make this go faster. Run `samtools mpileup` to see the usage information. Then run the basic command below. \n",
    "\n",
    "```\n",
    "samtools mpileup -r chr6:128405804-128605805 -f hg19.fa NA12878_child.sorted.bam > NA12878_child.mpileup\n",
    "```\n",
    "\n",
    "See what the pileup looks like with `head -n 100`. These files will be more convenient for SNP calling, since they \"pileup\" all the bases sequenced at each position in the reference genome. Read about pileup format here: http://samtools.sourceforge.net/pileup.shtml.\n",
    "\n",
    "<blockquote>\n",
    "**UNIX TIP**: Sometimes files with long lines are hard to read in the terminal screen because the lines get \"wrapped\". I like to pipe output to `less -S` which allows you to scroll horizontally. Try it out with the command below.\n",
    "</blockquote>\n",
    "\n",
    "```\n",
    "head -n 100 NA12878_child.mpileup | less -S\n",
    "```\n",
    "\n",
    "Now look at the usage of `samtools mpileup` to determine how to run the command on multiple BAM files at once (child, father, and mother, in that order so everyone's files are consistent). Output the results to `trio.mpileup`.\n",
    "\n",
    "To examine the output, we'll learn a new UNIX command, `grep`, which is a useful tool for searching files for specific string patterns (kind of like \"ctrl+F\" or \"find\" functions in other tools). Try the command below:\n",
    "\n",
    "```\n",
    "cat trio.mpileup | grep 128419307\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color=\"red\" size=\"3\" style=\"background-color:#f2e3e5\">**Question 1 (3 pts)**: How many reads does each smaple (child, mother, father) have covering the position chr6:128419307? What does it look like the genotype of each sample is? (TT, AT, or AA)? Answer the same questions for position chr6:128435303 (coverage and genotype for each sample).</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color=\"blue\" size=\"3\" style=\"background-color:#e2e6f1\">**Answer 1**</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Plotting the coverage distribution\n",
    "\n",
    "The mpileup format makes it straightforward to plot the coverage distribution for each sample. We can simply pull out the columns of the file for each sample correponding to coverage. \n",
    "\n",
    "<font color=\"red\" size=\"3\" style=\"background-color:#f2e3e5\">**Question 2 (4 pts)** Complete the code below to plot the distribution of coverage and compute the mean of the distribution for each sample. </font>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color=\"blue\" size=\"3\" style=\"background-color:#e2e6f1\">**Answer 2**</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Populating the interactive namespace from numpy and matplotlib\n",
      "Mean child coverage: 50.7579824201758\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjgAAADQCAYAAAAK/RswAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAHEZJREFUeJzt3X24pXVd7/H3Jx4NUUDIuAAdVIIzmiHOQUqPYZg8FUNlBicTDOWqwPRo1pilHKyTD5UPR6Qm4fBwTCTS46h4OIhwyAxkUIQYRKcRYwgBBZHiiIHf88f927AY9sPae/Zae62136/rWte+79/9W/f6rnu+M/Pd98Pvl6pCkiRpkvzQUgcgSZK02CxwJEnSxLHAkSRJE8cCR5IkTRwLHEmSNHEscCRJ0sSxwJEGKMnZSe5M8o8zbE+S9yXZmOT6JAcNO0apH+ayxo0FjjRY5wBHzLL9SGC/9joZOHMIMUkLcQ7mssaIBY40QFV1JXD3LF1WA+dV5ypglyR7Dic6qX/mssbNtksdwLDtvvvutWLFiqUOQyPk2muv/VZV7bFEH78XcGvP+ubWdvuWHZOcTPebMTvttNNzDzjggKEEqPFhLmtSLEYuL7sCZ8WKFaxfv36pw9AISfKNpY6hH1W1FlgLsGrVqjKPtSVzWZNiMXLZS1TS0roN2Kdnfe/WJo0bc1kjxQJHWlrrgFe0J1AOAe6tqsec0pfGgLmskbLsLlFJw5Tkw8ChwO5JNgNvBbYDqKq/AC4GjgI2AvcDr1yaSKXZmcsaNxY40gBV1fFzbC/glCGFIy2Yuaxx4yUqSZI0cSxwJEnSxPES1TK1Ys2n5v2eW95+9AAikSRp8XkGR5IkTRwLHEmSNHEscCRJ0sSxwJEkSRPHAkeSJE0cCxxJkjRxLHAkSdLEscCRJEkTxwJHkiRNHAscSZI0cZyqQX2b7/QOTu0gSVoqnsGRJEkTxwJHkiRNHAscSZI0cSxwJEnSxLHAkSRJE8cCR5IkTRwLHEmSNHEscCRJ0sSxwJEkSRPHAkcasCRHJLk5ycYka6bZ/pQklyf5UpLrkxy1FHFKczGXNU4scKQBSrINcAZwJLASOD7Jyi26/QFwYVU9BzgO+MBwo5TmZi5r3FjgSIN1MLCxqjZV1feBC4DVW/Qp4Alt+YnAvwwxPqlf5rLGigWONFh7Abf2rG9ubb1OA16eZDNwMfCa6XaU5OQk65Osv+uuuwYRqzQbc1ljZaAFTpJdklyU5CtJbkryk0l2S3Jpkq+1n7u2vknyvnZt9/okB/Xs54TW/2tJTuhpf26SG9p73pckg/w+0oAcD5xTVXsDRwHnJ3nM382qWltVq6pq1R577DH0IKU+mMsaGYM+g/Ne4H9X1QHATwA3AWuAy6pqP+Cytg7ddd392utk4EyAJLsBbwWeR3eK9K1TRVHr8+qe9x0x4O8jzddtwD4963u3tl4nARcCVNU/ADsCuw8lOql/5rLGysAKnCRPBF4InAVQVd+vqu/QXbM9t3U7Fzi2La8GzqvOVcAuSfYEDgcuraq7q+oe4FLgiLbtCVV1VVUVcF7PvqRRcQ2wX5J9k2xPd+Plui36/DNwGECS/0D3n4Ln7TVqzGWNlUGewdmXLrH/R3tk8INJdgKeXFW3tz7fBJ7clme6vjtb++Zp2h/D671aKlX1IHAqcAndGcwLq+rGJKcnOaZ1ewPw6iRfBj4MnNiKdmlkmMsaN9sOeN8HAa+pqquTvJdHLkcBUFWVZODJX1VrgbUAq1at8i+bhqqqLqa74bK37S09yxuA5w87Lmm+zGWNk0GewdkMbK6qq9v6RXQFzx3t8hLt551t+0zXd2dr33uadkmStMwNrMCpqm8CtybZvzUdBmygu2Y79STUCcDH2/I64BXtaapDgHvbpaxLgJck2bXdXPwS4JK27btJDmlPT72iZ1+SJGkZG+QlKujGQPhQuyFtE/BKuqLqwiQnAd8AXtb6Xkz3WOFG4P7Wl6q6O8nb6G5wAzi9qu5uy78FnAM8Dvh0e0mSpGVuoAVOVV0HrJpm02HT9C3glBn2czZw9jTt64FnbWWYkiRpwjiSsSRJmjgWOJIkaeJY4EiSpIljgSNJkiaOBY4kSZo4FjiSJGniWOBIkqSJY4EjSZImjgWOJEmaOBY4kiRp4ljgSJKkiWOBI0mSJs6cBU6SX06yc1v+gyQfTXLQ4EOTJElamH7O4PxhVd2X5AXAi4GzgDMHG5YkSdLC9VPgPNR+Hg2srapPAdsPLiRJkqSt00+Bc1uSvwR+Bbg4yQ59vk+SJGlJ9FOovAy4BDi8qr4D7Aa8caBRSZIkbYU5C5yquh/4OPBvSZ4CbAd8ZdCBSZIkLdS2c3VI8hrgrcAdwA9acwHPHmBckiRJC9bPJarXAvtX1TOr6sfby+JG6lOSI5LcnGRjkjUz9HlZkg1Jbkzy18OOUZqLeaxxM+cZHOBW4N5BByJNoiTbAGcAPwtsBq5Jsq6qNvT02Q94E/D8qronyY8sTbTS9MxjjaN+CpxNwBVJPgU8MNVYVX8+sKikyXEwsLGqNgEkuQBYDWzo6fNq4Iyqugegqu4cepTS7MxjjZ1+LlH9M3Ap3dg3O/e8JM1tL7qzoFM2t7ZePwb8WJK/T3JVkiOm21GSk5OsT7L+rrvuGlC40rQWLY/BXNZwzHkGp6r+K0CSx7f1fx10UNIysy2wH3AosDdwZZIfb8MyPKyq1gJrAVatWlXDDlKaQ195DOayhqOfuaieleRLwI3AjUmuTfLMwYcmTYTbgH161vdubb02A+uq6t+r6uvAV+n+o5BGhXmssdPPJaq1wOur6qlV9VTgDcBfDTYsaWJcA+yXZN8k2wPHAeu26PO/6H7rJcnudKf6Nw0zSGkO5rHGTj8Fzk5VdfnUSlVdAew0sIikCVJVDwKn0o0GfhNwYVXdmOT0JMe0bpcA306yAbgceGNVfXtpIpYeyzzWOOrrKaokfwic39ZfjlW51Lequhi4eIu2t/QsF/D69pJGknmscdPPGZxfB/YAPtpee7Q2SZKkkdTPU1T3AL89hFgkSZIWxYwFTpL3VNXrknyCbu6pR6mqY6Z5myRJ0pKb7QzO1D03fzqMQCRJkhbLjPfgVNW1bfHAqvq/vS/gwH4/IMk2Sb6U5JNtfd8kV7cJ2z7SHjkkyQ5tfWPbvqJnH29q7TcnObynfc7J3yRJ0vLTz03GJ0zTduI8PuO1dI8VTnkH8O6qegZwD3BSaz8JuKe1v7v1I8lKujEXngkcAXygFU1Tk78dCawEjm99JUnSMjdjgZPk+Hb/zb5J1vW8Lgfu7mfnSfYGjgY+2NYD/AxwUetyLnBsW17d1mnbD2v9VwMXVNUDbXTMjXQTvz08+VtVfR+YmvxNkiQtc7Pdg/N54HZgd+DPetrvA67vc//vAX6XRybnfBLwnTZoFDx6wraHJ3OrqgeT3Nv67wVc1bPP3vdsOfnb86YLIsnJwMkAT3nKU/oMXZIkjasZC5yq+gbwDeAnF7LjJD8H3FlV1yY5dGHhLQ4ndpMkaXmZ7THxz1XVC5Lcx6MfEw/doJVPmGPfzweOSXIUsCPwBOC9wC5Jtm1ncXonbJuazG1zkm2BJwLfZvZJ3uaa/E2SJC1Dsz1F9YL2c+eqekLPa+c+ihuq6k1VtXdVraC7SfizVfWrdHOUvLR1OwH4eFtexyM3NL+09a/Wflx7ympfutlpv0B/k79JkqRlaM6RjJM8HdhcVQ+0S03PBs6rqu8s8DN/D7ggyR8BXwLOau1nAecn2Uh3E/NxAG1CtwuBDcCDwClV9VCLbWryt22As6vqxgXGJEmSJkg/k23+LbAqyTPo7mP5OPDXwFH9fkibgfyKtryJ7gmoLft8D/jlGd7/x8AfT9P+mMnfJEmS+hkH5wftfplfAP57Vb0R2HOwYUmSJC1cPwXOvyc5nu7+mE+2tu0GF5IkSdLW6afAeSXdo+J/XFVfbzf6nj/HeyRJkpbMnAVOVW0Afge4Icmz6G44fsfAI5MkSVqgfp6iOpRuCoVb6MbA2SfJCVV15WBDkyRJWph+nqL6M+AlVXUzQJIfAz4MPHeQgUmSJC1UP/fgbDdV3ABU1VfxJmNJkjTC+jmDsz7JB4H/2dZ/FVg/uJAkSZK2Tj8Fzm8CpwC/3db/DvjAwCKSJEnaSnMWOG2KhvcDlwE/AG6uqu8PPDJJkqQFmvMenCRHA/9ENxP4+4GNSY4cdGDSpEhyRJKbk2xMsmaWfr+UpJKsGmZ8Ur/MZY2Tfp+ielFVbYSHJ9/8FPDpQQYmTYIk2wBnAD8LbAauSbKujS/V229n4LXA1cOPUpqbuaxx089TVPdNFTfNJuC+AcUjTZqDgY1Vtald2r0AWD1Nv7cB7wC+N8zgpHkwlzVW+ilw1ie5OMmJSU4APkFXuf9ikl8ccHzSuNsLuLVnfXNre1iSg4B9qupTs+0oyclJ1idZf9dddy1+pNLszGWNlX4KnB2BO4CfBg4F7gIeB/w88HMDi0xaBpL8EPDnwBvm6ltVa6tqVVWt2mOPPQYfnDQP5rJGTT9PUb1yGIFIE+o2YJ+e9b1b25SdgWcBVyQB+FFgXZJjqsrxpjRKzGWNlX7O4EhauGuA/ZLsm2R74Dhg3dTGqrq3qnavqhVVtQK4CvA/BI0ic1ljxQJHGqCqehA4FbgEuAm4sKpuTHJ6kmOWNjqpf+ayxk0/j4lL2gpVdTFw8RZtb5mh76HDiElaCHNZ46Sfgf7+oGd5h8GGI0mStPVmPIOT5PeAK4GXAn/Umv8BOGgIcWmeVqyZ9alMSZKWldkuUX0F+GXgaUn+rq0/Kcn+VXXzUKKTJElagNkuUX0H+H1gI934N+9t7WuSfH7AcUmSJC3YbGdwDgfeAjydbvCm64F/c1wcSZI06mY8g1NVv19VhwG3AOcD2wB7JPlckk8MKT5JkqR56+cx8UvaQE3rk/xmVb0gye6DDkySJGmh5nxMvKp+t2f1xNb2rUEFJEmStLXmNZJxVX15UIFIkiQtFqdqkCRJE8cCR5IkTRwLHEmSNHEscCRJ0sQZ2GziSfYBzgOeDBSwtqrem2Q34CPACroxdl5WVfckCd1oyUcB9wMnVtUX275OAKYm/fyjqjq3tT8XOAd4HN0Mt6+tqhrUd9L8zHd+rFvefvSAIpEkLTeDPIPzIPCGqloJHAKckmQlsAa4rKr2Ay5r6wBHAvu118nAmQCtIHor8DzgYOCtSXZt7zkTeHXP+44Y4PeRJEljYmAFTlXdPnUGpqruA24C9gJWA+e2bucCx7bl1cB51bkK2CXJnnRTRlxaVXdX1T3ApcARbdsTquqqdtbmvJ59SZKkZWwo9+AkWQE8B7gaeHJV3d42fZPuEhZ0xc+tPW/b3Npma988TbskSVrmBl7gJHk88LfA66rqu73b2pmXgd8zk+TkJOuTrL/rrrsG/XGSJGmJDbTASbIdXXHzoar6aGu+o11eov28s7XfBuzT8/a9W9ts7XtP0/4YVbW2qlZV1ao99thj676UJEkaeQMrcNpTUWcBN1XVn/dsWgec0JZPAD7e0/6KdA4B7m2Xsi4BXpJk13Zz8UvoJgC9HfhukkPaZ72iZ1+SJGkZG9hj4sDzgV8DbkhyXWv7feDtwIVJTgK+AbysbbuY7hHxjXSPib8SoKruTvI24JrW7/Squrst/xaPPCb+6faSJEnL3MAKnKr6HJAZNh82Tf8CTplhX2cDZ0/Tvh541laEKUmSJpAjGUuSpIljgSMNWJIjktycZGOSNdNsf32SDUmuT3JZkqcuRZzSbMxjjRsLHGmAkmwDnEE3UvdK4Pg2onevLwGrqurZwEXAO4cbpTQ781jjyAJHGqyDgY1Vtamqvg9cQDdq98Oq6vKqur+tXsWjhz+QRoF5rLFjgSMN1kwjcc/kJGZ4GtABK7WEFi2PwVzWcFjgSCMiycuBVcC7ptvugJUaB3PlMZjLGo5BjoMjaeaRuB8lyYuBNwM/XVUPDCk2qV/mscaOZ3CkwboG2C/Jvkm2B46jG7X7YUmeA/wlcExV3TnNPqSlZh5r7FjgSANUVQ8Cp9JNOXITcGFV3Zjk9CTHtG7vAh4P/E2S65Ksm2F30pIwjzWOvEQlDVhVXUw3FUlv21t6ll889KCkeTKPNW48gyNJkiaOBY4kSZo4FjiSJGniWOBIkqSJY4EjSZImjgWOJEmaOBY4kiRp4ljgSJKkiWOBI0mSJo4FjiRJmjgWOJIkaeJY4EiSpIljgSNJkiaOBY4kSZo4FjiSJGniWOBIkqSJs+1SB6DprVjzqaUOQZKkseUZHEmSNHEscCRJ0sTxEpVGxnwvy93y9qMHFIkkadx5BkeSJE0cCxxJkjRxLHAkSdLEGfsCJ8kRSW5OsjHJmqWOR9rSXDmaZIckH2nbr06yYvhRSnMzlzVOxrrASbINcAZwJLASOD7JyqWNSnpEnzl6EnBPVT0DeDfwjuFGKc3NXNa4GfenqA4GNlbVJoAkFwCrgQ2D/mAH4lOf+snR1cBpbfki4P1JUlU1zEClOZjLGivjXuDsBdzas74ZeN6WnZKcDJzcVv81yc3T7Gt34FuLHuHCGMvMHo4ni/e74VMXbU+P1U+OPtynqh5Mci/wJLY47lvk8QNJ/nEgEc/PKOTHKMQAoxHH/gPc9yTn8ij82cFoxDEKMcAi5PK4Fzh9qaq1wNrZ+iRZX1WrhhTSrIxlZqMWzzD15vGoHIdRiGMUYhiVOJKsX8rP79eo5fIoxDAqcYxCDFNxbO0+xvoeHOA2YJ+e9b1bmzQq+snRh/sk2RZ4IvDtoUQn9c9c1lgZ9wLnGmC/JPsm2R44Dli3xDFJvfrJ0XXACW35pcBnvWdBI8hc1lgZ60tU7RrvqcAlwDbA2VV14wJ3N+slrCEzlpmNWjyzmilHk5wOrK+qdcBZwPlJNgJ30/3HMZdROQ6jEMcoxACjEcfAYpjwXB6FGGA04hiFGGAR4ojFtSRJmjTjfolKkiTpMSxwJEnSxFn2Bc5STvWQZJ8klyfZkOTGJK9t7acluS3Jde111BBjuiXJDe1z17e23ZJcmuRr7eeuQ4hj/57vf12S7yZ53VIem2HZmuHwk7yptd+c5PABxvD6lrfXJ7ksyVN7tj3U8+ezVTf99xHHiUnu6vm8V/VsO6Hl7NeSnLDlexcxhnf3fP5Xk3ynZ9uiHIskZye5c6bxYtJ5X4vx+iQH9WxblOOwgJiXPI/7jGPguTwKedxnHJOVy1W1bF90N8r9E/A0YHvgy8DKIX7+nsBBbXln4Kt0Q6CfBvzOEh2TW4Ddt2h7J7CmLa8B3rEEf07fpBuQb8mOzRC/66w5CfwW8Bdt+TjgI215Zeu/A7Bv2882A4rhRcAPt+XfnIqhrf/rEI/FicD7p3nvbsCm9nPXtrzrIGLYov9r6G6+Xexj8ULgIOAfZ9h+FPBpIMAhwNWLeRzGMY9HJZdHIY+Xay4v9zM4Dw89XlXfB6aGHh+Kqrq9qr7Ylu8DbqIbCXTUrAbObcvnAscO+fMPA/6pqr4x5M9dCv3kZO+fx0XAYUnS2i+oqgeq6uvAxra/RY+hqi6vqvvb6lV0Y6Istq35+3k4cGlV3V1V9wCXAkcMIYbjgQ8v4HNmVVVX0j2VNJPVwHnVuQrYJcmeLN5xmK9RyOO+4hhCLo9CHi8kjrHP5eVe4Ew39PiSFBjt9OxzgKtb06nt9NzZGcIloR4F/J8k16YbTh3gyVV1e1v+JvDkIcYD3W93vX/RlurYDEM/Ofmo4fCBqeHwFyuf57ufk+h+45qyY5L1Sa5KsjXFcL9x/FLLh4uSTA1EN/Rj0S5t7At8tqd5sY7FXGaKc6n+jRuFPO43jl6DyOVRyON57WtScnm5FzgjIcnjgb8FXldV3wXOBJ4OHAjcDvzZEMN5QVUdRDdj8ClJXti7sbpzhUMbWyDdgGLHAH/Tmpby2GgLSV4OrALe1dP81OqGev/PwHuSPH2AIXwCWFFVz6b7je7cOfoP0nHARVX1UE/bMI+FtsIS5/Io5TFMSC4v9wJnyad6SLIdXXHzoar6KEBV3VFVD1XVD4C/YuGnZ+etqm5rP+8EPtY++452ipD2885hxUNXaH2xqu5ocS3ZsRmSrRkOf7Hyua/9JHkx8GbgmKp6YKq9J4c2AVfQnZlciDnjqKpv93z2B4Hnzuc7LEYMPbY807iYx2IuM8W5VP/GjUIe9xvHoHN5FPJ4vvuajFxejJuGxvVFN5LzJrpTcVM3XT1ziJ8f4DzgPVu079mz/F/orkcPI56dgJ17lj9Pd43zXTz6JuN3DvEYXQC8cqmPzRC/75w5CZzCo2/OvLAtP5NH35y5iYXdZNxPDM+hu2Fxvy3adwV2aMu7A19jgTfu9xlHbz78AnBVW94N+HqLZ9e2vNsgYmj9DqC7QT+DOBZtHyuY+cbMo3n0jZlfWMzjMI55PCq5PAp5vFxzeaBJPg4vuju2v9oS/M1D/uwX0F3uuR64rr2OAs4Hbmjt63qTf8DxPK0l/ZeBG6eOB9118ctaUn9mGP9Ats/die43uif2tC3JsRlyXjwmJ4HT6X67BNiR7pLdRuALwNN63vvm9r6bgSMHGMNngDt68nZda/+p9ufz5fbzpAEfiz9pufpl4HLggJ73/no7RhvpKZIXO4a2fhrw9i3et2jHgu636duBf6e79+Ak4DeA32jbA5zRYrwBWLXYx2Ec83hUcnkU8ng55rJTNUiSpImz3O/BkSRJE8gCR5IkTRwLHEmSNHEscCRJ0sSxwJEkSRPHAmcJJPmTJC9KcmySNw3pM89J8tIB7fvQJJ+cYdsHk6xsy7ck2X2aPqcl+Z1BxCZJWp4scJbG8+gmdftp4Mq5OrdRPsdSVb2qqjYsdRySpOXFAmeIkrwryfXAfwT+AXgVcGaSt0zT95wkf5HkauCdSXZqk0t+IcmXkqxu/VYk+bskX2yvn2rtSfL+JDcn+QzwIz37fnuSDW1itz+d53d4RpLPJPly+7yp+Uge3yaJ+0qSD7VZgUlyRZJV0+znzUm+muRzwP7ziUGSpLmM7ZmBcVRVb0xyIfAK4PXAFVX1/FnesjfwU1X1UJL/Bny2qn49yS7AF1rhcifws1X1vST70Y0SuYpuuO/9gZV0s39vAM5O8qS27YCqqrav+fgQ3SiXH0uyI12RvA/dcOfPBP4F+Hvg+cDnpttBkufSDc1+IF0OfhG4dp5xSJI0I8/gDN9BdMNdHwDcNEffv6lHZnN9CbAmyXV0E53tCDwF2A74qyQ30A17vrL1fyHw4eompvwXHpn2/l7ge8BZSX4RuL/fwJPsDOxVVR8DqKrvVdXU+79QVZurmwTzOrq5Rmbyn4CPVdX91c2evq7fGCRJ6odncIYkyYHAOXRnZb4F/HDXnOuAn6yq/zfN2/6tdxfAL1XVzVvs9zS6eVR+gq5g/d5scVTVg0kOBg4DXgqcCvzMFvu8hO6sz/qqelWfX/GBnuWHMLckSUvIMzhDUlXXVdWBdBOdraQ7o3J4VR04Q3GzpUuA1/Tc2zI1Vf0TgdvbmZNfA7Zp7VcCv5JkmyR7Ai9q73s83eSVF9PNxv0T08Q6Fdertmi/D9ic5Ni2rx2S/PA8DsOUK4FjkzyunRX6+QXsQ5KkGVngDFGSPYB7WjFywDyfLnob3eWo65Pc2NYBPgCckGTqstfUWZ+P0c3+vQE4j+6mZoCdgU+2m50/R3cv0Hz8GvDb7f2fB350nu+nqr4IfITuUt2ngWvmuw9JkmbjbOKSJGnieAZHkiRNHAscSZI0cSxwJEnSxLHAkSRJE8cCR5IkTRwLHEmSNHEscCRJ0sT5/9WAJrfYy8W4AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 576x216 with 3 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "%pylab inline\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "covdata = pd.read_csv(os.path.join(HOMEDIR, \"trio.mpileup\"), sep=\"\\t\",\n",
    "                     names = [\"cov.child\", \"cov.father\", \"cov.mother\"],\n",
    "                     usecols = [3, 6, 9])\n",
    "\n",
    "# Set up the plot\n",
    "fig = plt.figure()\n",
    "fig.set_size_inches((8, 3))\n",
    "\n",
    "# Plot child coverage distribution\n",
    "ax = fig.add_subplot(131)\n",
    "ax.hist(covdata[\"cov.child\"])\n",
    "ax.set_xlabel(\"# reads - child\")\n",
    "ax.set_ylabel(\"# positions\")\n",
    "print(\"Mean child coverage: %s\"%np.mean(covdata[\"cov.child\"]))\n",
    "\n",
    "# Plot father coverage distribution\n",
    "ax = fig.add_subplot(132)\n",
    "# TODO fill in this section\n",
    "\n",
    "# Plot mother coverage distribution\n",
    "ax = fig.add_subplot(133)\n",
    "# TODO fill in this section\n",
    "\n",
    "# Clean up the plot spacing\n",
    "fig.tight_layout();"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Running a SNP caller (Varscan)\n",
    "\n",
    "Now we will run a SNP caller to identify variants in our three samples. We could parse the mpileup file ourselves to do this, but there are already dozens of existing SNP callers out there so for now we'll choose not to reinvent the wheel.\n",
    "\n",
    "Until now, all the programs you have been using (BWA, fastqc, samtools...) were pre-installed for you in our public folder. Here you’ll get to try your hand at installing one type of program file.\n",
    "\n",
    "Use the command below (`curl` - it's a linux/unix command to access websites) to download VarScan into your `week1` folder. Notice that we are redirecting the contents of the download into a .jar file.\n",
    "\n",
    "```\n",
    "curl -L https://sourceforge.net/projects/varscan/files/VarScan.v2.3.9.jar/download > VarScan.jar\n",
    "```\n",
    "\n",
    "Try to bring up the man page with:\n",
    "```\n",
    "java -jar VarScan.jar\n",
    "```\n",
    "\n",
    "This install method won’t be the same for all bioinformatics programs, and it might not even be the same for other java programs, as installing open source software often has dependencies (prerequisite software) and can be sensitive to where its installed and how it is executed. We'll try to touch on this throughout the course.\n",
    "\n",
    "\n",
    "On the VarScan man page, you should see several commands. We are interested in the mpileup2snp command. Go ahead and enter it with `-h` to bring up the manual.\n",
    "\n",
    "```\n",
    "java -jar VarScan.jar mpileup2snp -h\n",
    "```\n",
    "\n",
    "There are lots of ways we could filter our data. VarScan lets the user define their own cutoffs for including data in calling variants, other programs rely on complex statistical procedures to evaluate the likelihood of real mutations.\n",
    "\n",
    "The first option we are interested in setting today is the -`-min-var-frequency` option. This sets the minimum % of non-reference bases at a position required to call it a variant in the sample. Set this too low, and we'll pick up mostly sequencing errors. Set this too high, and we'll miss real variants. We're also interested in `--min-freq-for-hom`, the minimum % of non-reference bases to call a variant homozygous.\n",
    "\n",
    "Choose reasonable thresholds that will allow us to pick up both heterozygous and homozygous variants. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color=\"red\" size=\"3\" style=\"background-color:#f2e3e5\">**Question 3 (4 pts)**: Choose reasonable thresholds to modify and run the command below to perform SNP calling.</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color=\"blue\" size=\"3\" style=\"background-color:#e2e6f1\">**Answer 3**</font>\n",
    "\n",
    "```\n",
    "java -jar VarScan.jar mpileup2snp \\\n",
    "  trio.mpileup \\\n",
    "  --min-var-frequency N \\\n",
    "  --min-freq-for-hom N \\\n",
    "  --variants \\\n",
    "  --p-value 0.01 \\\n",
    "  --output-vcf 1 > trio.vcf\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Intro to VCF files\n",
    "\n",
    "VarScan outputs variants in VCF (variant call format), a standard genomics format. To read about all the gory details of VCF files, see the specs: https://samtools.github.io/hts-specs/VCFv4.2.pdf\n",
    "\n",
    "First take a peak and scroll through the file:\n",
    "\n",
    "```\n",
    "cat trio.vcf | less -S\n",
    "```\n",
    "\n",
    "Key things to notice:\n",
    "\n",
    "* The header lines (starting with \"#\") give general information about the file such as the tool use to generate it and a description of all the different data fields in the file.\n",
    "* Below the header, there is one line per variant. For each variant, you can find the chromosome, position, reference allele, alternate allele, and additional info. Following these fields are the variant call and statistics for each sample\n",
    "* \"INFO\" fields describe per-variant statistics. \"FORMAT\" fields describe per-sample statistics at each variant.\n",
    "\n",
    "See lecture slides for more VCF info."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color=\"red\" size=\"3\" style=\"background-color:#f2e3e5\">**Question 4 (4 pts)**: Look at the position chr6:128593590 (it might be helpful to use grep for this). What is the genotype for each of the three samples? What is the p-value for each?</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color=\"blue\" size=\"3\" style=\"background-color:#e2e6f1\">**Answer 4**</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Inspecting putative variants\n",
    "\n",
    "It is always a good idea to look at the raw data used to make genotype calls. Use `samtools tview` to navigate to position chr6:128593590 in the BAM files for each of your samples. Does the raw data support the genotype called by VarScan? Look at the alignments for several other positions with SNP calls."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Mutation identification and annotation\n",
    "\n",
    "We'd now like to identify potential *de novo* mutations that occurred in the child genome but that were not passed down by either parent. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color=\"red\" size=\"3\" style=\"background-color:#f2e3e5\">**Question 5 (15 pts)** In the space below, write a small program that identifies candidate mutations in the child genome.\n",
    "\n",
    "You may do this using whatever means you want, but example template code to get you started using the PyVCF library is provided. This can also be done using bash commands. Your code must run from within the notebook and output a list of candidate mutation positions.\n",
    "\n",
    "Note you could also do this manually, since our VCF file is pretty small. But we will give only 1 point for solutions with no code provided.\n",
    "\n",
    "<br><br>\n",
    "Your code should output one line per candidate mutation in the format:<br>\n",
    "chrom,position,child genotype;father genotype;mother genotype\n",
    "</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color=\"blue\" size=\"3\" style=\"background-color:#e2e6f1\">**Answer 5**</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr6:128414945:T/T;T/T;T/T\n",
      "chr6:128419307:T/A;T/A;T/A\n",
      "chr6:128421762:T/T;T/T;T/T\n",
      "chr6:128435303:C/T;C/T;C/T\n",
      "chr6:128442104:G/G;G/G;G/T\n",
      "chr6:128446509:T/T;T/T;T/T\n",
      "chr6:128450019:T/G;T/G;T/T\n",
      "chr6:128465668:C/T;C/C;C/T\n",
      "chr6:128469937:T/A;T/A;T/A\n",
      "chr6:128471913:A/C;A/C;A/C\n",
      "chr6:128484504:A/G;A/G;A/A\n",
      "chr6:128486139:A/C;A/C;A/C\n",
      "chr6:128488195:T/C;T/C;T/C\n",
      "chr6:128493991:G/T;G/T;G/T\n",
      "chr6:128494093:C/G;C/G;C/G\n",
      "chr6:128496528:T/T;T/T;T/C\n",
      "chr6:128496557:T/T;T/T;T/T\n",
      "chr6:128503550:T/C;T/C;T/T\n",
      "chr6:128505804:A/C;A/A;A/A\n",
      "chr6:128514053:G/A;G/A;G/A\n",
      "chr6:128515718:T/C;T/C;T/T\n",
      "chr6:128517169:T/T;T/T;T/C\n",
      "chr6:128532763:G/C;G/C;G/C\n",
      "chr6:128559329:C/T;C/T;C/T\n",
      "chr6:128564669:G/T;G/T;G/T\n",
      "chr6:128568999:C/T;C/T;C/C\n",
      "chr6:128572877:C/A;C/A;C/A\n",
      "chr6:128583972:G/A;G/G;G/A\n",
      "chr6:128583983:T/G;T/T;T/G\n",
      "chr6:128585618:A/C;A/C;A/C\n",
      "chr6:128591818:T/G;T/G;G/G\n",
      "chr6:128593590:T/T;T/T;T/G\n",
      "chr6:128599580:C/A;C/A;C/A\n",
      "chr6:128599902:A/A;A/A;A/A\n",
      "chr6:128602090:A/C;A/C;A/C\n",
      "chr6:128602577:T/C;T/C;T/C\n",
      "chr6:128602796:C/T;C/T;C/T\n",
      "chr6:128603361:C/T;C/T;C/T\n",
      "chr6:128604078:C/T;C/T;C/T\n",
      "chr6:128604430:G/A;G/A;G/A\n",
      "chr6:128604571:A/C;A/C;A/C\n"
     ]
    }
   ],
   "source": [
    "#### TODO replace this with your answer ####\n",
    "#### Template code is provided to get you started with the PyVCF library ####\n",
    "#### But you DO NOT have to use it! ####\n",
    "#### For more info see https://github.com/jamescasbon/PyVCF\n",
    "\n",
    "# This code simply prints out each position in the VCF file and the genotypes\n",
    "# of each sample. Modify to print out only positions with putative mutations\n",
    "# in the child.\n",
    "\n",
    "import vcf\n",
    "\n",
    "reader = vcf.Reader(open(os.path.join(HOMEDIR, \"trio.vcf\"), \"r\"))\n",
    "\n",
    "for record in reader:\n",
    "    chrom  = record.CHROM\n",
    "    pos = record.POS\n",
    "    child_gt = record.genotype(\"Sample1\").gt_bases\n",
    "    father_gt = record.genotype(\"Sample2\").gt_bases\n",
    "    mother_gt = record.genotype(\"Sample3\").gt_bases\n",
    "\n",
    "    print(\"%s:%s:%s;%s;%s\"%(chrom, pos, child_gt, father_gt, mother_gt))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's figure out if the mutations identified are likely to have an impact on any particular genes. We'll use the UCSC Genome Browser (https://genome.ucsc.edu/) to help with this. Go to the genome browser home page. Then on the top menu bar, hover over \"Genomes\" and choose the hg19 reference genome.\n",
    "\n",
    "<font color=\"red\" size=\"3\" style=\"background-color:#f2e3e5\">**Question 6 (5 pts)**: In the search bar, navigate to the position of each mutation. e.g. \"chr6:XXXX\" in the search bar. Does your mutation fall in any genes? If yes, determine whether it is predicted to change the amino acid structure (i.e. is this a synonymous variant? missense? nonsense?) of the protein and describe the change below.</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color=\"blue\" size=\"3\" style=\"background-color:#e2e6f1\">**Answer 6**</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Additional challenge questions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "TODO: anomalies in sequencing coverage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "TODO: anomalies in insert size distributions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "TODO: SNP calling error sources"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
